tests/e2e/tests/pilot/performance/serviceentry_test.go

```diff
@@ -0,0 +1,328 @@
+	pilotAddr := fmt.Sprintf("127.0.0.1:%d", pilotGrpcPort)
+
+	tearDown := initLocalPilotTestEnv(t, mcpServer.Port, pilotAddr, debugAddr)
+	defer tearDown()
+
+	t.Log("run sidecar envoy(s)...")
+	adscs := adsConnectAndWait(*envoyCount, pilotAddr)
+	for _, adsc := range adscs {
+		defer adsc.Close()
+	}
+
+	t.Log("check for service entries in pilot's debug endpoint...")
+	g.Eventually(func() (int, error) {
+		return registeredServiceEntries(fmt.Sprintf("http://127.0.0.1:%d/debug/configz", pilotDebugPort))
+	}, "30s", "1s").Should(gomega.Equal(*serviceEntryCount))
+
+	t.Log("check for registered endpoints in envoys's debug endpoint...")
+	g.Eventually(func() error {
+		for _, adsc := range adscs {
+			var data map[string]interface{}
+			err = json.Unmarshal([]byte(adsc.EndpointsJSON()), &data)
+			if err != nil {
+				return err
+			}
+			if len(data) != *serviceEntryCount {
+				return errors.New("endpoint not registered")
+			}
+		}
+		return nil
+	}, "30s", "1s").ShouldNot(gomega.HaveOccurred())
+
+	terminate()
+}
+
+func runSnapshot(mcpServer *mcptest.Server) {
+	configInterval := time.NewTicker(time.Second * 3)
+	v := 0
+	b := snapshot.NewInMemoryBuilder()
+	for {
+		select {
+		case <-configInterval.C:
+			for {
+				// wait for pilot to make a MCP request
+				if status := mcpServer.Cache.Status(snapshot.DefaultGroup); status != nil {
+					if status.Watches() > 0 {
+						break
+					}
+				}
+				time.Sleep(10 * time.Millisecond)
+			}
+			v++
+			version := strconv.Itoa(v)
+			for _, m := range model.IstioConfigTypes {
+				if m.MessageName == model.ServiceEntry.MessageName {
+					b.Set("type.googleapis.com/istio.networking.v1alpha3.ServiceEntry", version, generateServiceEntries())
+				} else if m.MessageName == model.Gateway.MessageName {
+					b.Set("type.googleapis.com/istio.networking.v1alpha3.Gateway", version, generateGateway())
+				} else {
+					b.Set(fmt.Sprintf("type.googleapis.com/%s", m.MessageName), version, []*mcp.Envelope{})
+				}
+			}
+
+			lastSnapshot := time.Now()
+			shot := b.Build()
+			mcpServer.Cache.SetSnapshot(snapshot.DefaultGroup, shot)
+			b = shot.Builder()
+
+			// wait for client to ACK the pushed snapshot
+			for {
+				if status := mcpServer.Cache.Status(snapshot.DefaultGroup); status != nil {
+					if status.LastWatchRequestTime().After(lastSnapshot) {
+						break
+					}
+				}
+				time.Sleep(10 * time.Millisecond)
+			}
+		}
+	}
+}
+
+func terminate() {
+	<-time.After(time.Second * time.Duration(*terminateAfter))
+}
+
+func adsConnectAndWait(n int, pilotAddr string) (adscs []*adsc.ADSC) {
+	for i := 0; i < n; i++ {
+		c, err := adsc.Dial(pilotAddr, "", &adsc.Config{
+			IP: net.IPv4(10, 10, byte(i/256), byte(i%256)).String(),
+		})
+		if err != nil {
+			log.Fatalf("Error connecting %s\n", err)
+		}
+		c.Watch()
+		_, err = c.Wait("rds", 10*time.Second)
+		if err != nil {
+			log.Fatalf("Error getting initial config %s\n", err)
+		}
+
+		if len(c.EDS) == 0 {
+			log.Fatal("No endpoints")
```

Comment:
These `log.Fatal` should either be `t.Fatal` or the error should be returned to the caller which can invoke `t.Fatal`.